lib/Dialect/Comb/CombFolds.cpp

```diff
@@ -73,6 +74,128 @@ static bool tryFlatteningOperands(Op op, PatternRewriter &rewriter) {
+
+  for (auto *user : users) {
+    if (auto extractOp = dyn_cast<ExtractOp>(user)) {
+      size_t lowBit = extractOp.lowBit();
+      size_t highBit = extractOp.getType().getWidth() + lowBit - 1;
+      highestBitRequired = std::max(highestBitRequired, highBit);
+      lowestBitRequired = std::min(lowestBitRequired, lowBit);
+      continue;
+    }
+
+    highestBitRequired = originalOpWidth - 1;
+    lowestBitRequired = 0;
+    break;
+  }
+
+  return {lowestBitRequired, highestBitRequired};
+}
+
+// narrowOperationWidth(...) attempts to create a new instance of Op
+// with createOp using narrowed operands. Ie: transforming f(x, y) into
+// f(x[n:m], y[n:m]), It analyzes all usage sites of narrowingCandidate and
+// mutate them to the newly created narrowed version if it's benefitial.
+//
+// narrowTrailingBits determines whether undemanded trailing bits should
+// be aggressively stripped off.
+//
+// This function requires the narrowingCandidate to have at least 1 use.
+//
+// Returns true if IR is mutated. IR is mutated iff narrowing happens.
+static bool
+narrowOperationWidth(Operation *narrowingCandidate, ValueRange inputs,
+                     bool narrowTrailingBits, PatternRewriter &rewriter,
+                     std::function<Value(ArrayRef<Value>)> createOp) {
+  // If the result is never used, no point optimizing this. It will
+  // also complicated error handling in getLowestBitAndHigestBitRequired.
+  auto usersIterator = narrowingCandidate->getUsers();
+  assert(!usersIterator.empty() &&
+         "narrowingCandidate must have at least one use.");
+  assert(narrowingCandidate->getNumResults() == 1 &&
+         "narrowingCandidate must have exactly one result");
+  IntegerType narrowingCandidateType =
+      narrowingCandidate->getResultTypes().front().cast<IntegerType>();
+
+  size_t highestBitRequired;
+  size_t lowestBitRequired;
+  size_t originalOpWidth = narrowingCandidateType.getIntOrFloatBitWidth();
+
+  std::tie(lowestBitRequired, highestBitRequired) =
+      getLowestBitAndHighestBitRequired(narrowingCandidate, narrowTrailingBits,
+                                        originalOpWidth);
+
+  // Give up, because we can't make it narrower than it already is.
+  if (lowestBitRequired == 0 && highestBitRequired == originalOpWidth - 1)
+    return false;
+
+  auto loc = narrowingCandidate->getLoc();
+  size_t narrowedWidth = highestBitRequired - lowestBitRequired + 1;
+  auto narrowedType = rewriter.getIntegerType(narrowedWidth);
+  Value narrowedOperation =
+      createOp(SmallVector<Value>(llvm::map_range(inputs, [&](auto input) {
+        return rewriter.create<ExtractOp>(loc, narrowedType, input,
+                                          lowestBitRequired);
+      })));
+
+  // Replace all the use-site's extract operation with the newly minted narrowed
+  // version.
+  SmallVector<Operation *, 16> users(usersIterator.begin(),
+                                     usersIterator.end());
+  for (auto *user : users) {
```

Comment:
Does it work to do:

```
while (!narrowingCandidate->use_empty()) {
  auto *user = *narrowingCandidate->user_begin();
```

If this doesn't work for some reason, please sink the definition of `usesIterator` down here and pass it directly into the smallvector ctor

Reply:
There is also `llvm::make_early_inc_range`, which I've used for this kind of thing: https://github.com/llvm/circt/blob/main/lib/Conversion/HWToLLHD/HWToLLHD.cpp#L315

Reply:
I didn't know about `llvm::make_early_inc_range` -- thanks showing it!